pkg/pool-weighted/contracts/WeightedMath.sol

```diff
@@ -214,6 +214,33 @@ contract WeightedMath {
         return nonTaxableAmount.add(taxableAmount.divUp(FixedPoint.ONE.sub(swapFeePercentage)));
     }
 
+    function _calcTokensInGivenExactBptOut(
+        uint256[] memory balances,
+        uint256 bptAmountOut,
+        uint256 bptTotalSupply
+    ) internal pure returns (uint256[] memory amountsIn) {
+        /*************************************************************************************************
+        // tokensInForExactBptOut                                                                       //
+        // (per token)                                                                                  //
+        // aI = amountIn                   /  totalBPT + bptOut      \                                  //
+        // b = balance           aI = b * | --------------------- - 1 |                                 //
+        // bptOut = bptAmountOut           \       totalBPT          /                                  //
+        // bpt = totalBPT                                                                               //
+        *************************************************************************************************/
+
+        // Tokens in, so we round up overall.
+
+        amountsIn = new uint256[](balances.length);
+
+        // Calculate the factor by which the invariant will increase after minting BPTAmountOut
+        uint256 balanceMultiplier = bptTotalSupply.add(bptAmountOut).divUp(bptTotalSupply).sub(FixedPoint.ONE);
+
+        // Balances need to increase by (invariantRatio - 1)
+        for (uint256 i = 0; i < balances.length; i++) {
+            amountsIn[i] = balanceMultiplier.mulUp(balances[i]);
+        }
+    }
```

Comment:
This is needlessly complex:

`(total + out) / total - 1` equals 
`total / total + out / total - 1`, which equals 
`1 + out / total - 1`, which of course equals
`out / total`

Reply:
In order words, the amounts in are just the bptout/total ratio of the current balances. I'd copy the exit implementation:

```
        uint256 bptRatio = bptAmountIn.divDown(totalBPT);

        uint256[] memory amountsOut = new uint256[](balances.length);
        for (uint256 i = 0; i < balances.length; i++) {
            amountsOut[i] = balances[i].mulDown(bptRatio);
        }

        return amountsOut;
```

Reply:
Too late at night :) Now in the proportional PR